python: Refine highlighting

JaagupAverin opened the pull request:
Fixes:
   * Types in binary unions as per [PEP 604](https://peps.python.org/pep-0604/) not highlighted;
   * `except*` keyword not highlighted;
   * Classes beginning with `_` not recognized as such, however `_` is a valid first character for private classes; additionally the regex for parsing constant/class names appeared inconsistent and incomplete so was adjusted;
   * Builtin types such as `float`, `dict`, etc not recognized as types;
   * **Update:** decorators with arguments not recognized as decorators;
   * **Update:** docstrings after type alias assignments not recognized as docstrings;
   * **Update:** `and/in/is/not/or/is not/not in` not capturable as keywords;
   * **Update:** decorators with "nesting" (@x.y.z) not recognized as decorators;

Before:
![new_before](https://github.com/user-attachments/assets/6f05262e-be3b-41bf-aee6-26438c2bf254)

After:
![new_after](https://github.com/user-attachments/assets/408c481c-5eb9-40c9-8e18-52ebf5a121d3)

Release Notes:

- N/A

JaagupAverin committed:
python: refine highlighting

JaagupAverin commented on crates/languages/src/python/highlights.scm:
```diff
@@ -1,18 +1,28 @@
+; Any identifiers within a type are in practice going to be type identifiers; this nested scan
+; is currently the only way to support cases such as `type MyList[T] = T | list[T]`
+(type (_ (identifier) @type))
+(type (_ (_ (identifier) @type)))
```
@osiewicz Currently this PR is still in draft, but already in advance I feel that this approach here is incredibly silly. I am not sure if it is possible to apply a rule for an arbitrarily nested case such as this?
```python
type AnyTasks[T: ExecutableTask | Task] = T | _TaskList[T] | list[T | _TaskList[T]] | None
```
I feel like the problem boils down to not getting adequate output from the parser so we're having to do more complicated stuff in the highlighter.

osiewicz replied:
Hmm. I'll revisit your PR tomorrow, but off the top of my head that's not going to be easy to pull off cleanly. Tree-sitter doesn't support matches of arbitrary depth. 
Besides, I think we should fall back on [semantic highlighting](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) for cases like these. While in principle the only valid kind of identifier in that position is a type, users code might be invalid. Thus, if possible, I'd rather have us tackle semantic tokens instead of handling these cases at the tree-sitter level.

Don't get me wrong, I really appreciate your recent efforts on Python highlighting. It's just that this PR runs into a wall that should be brought down, not chipped away at :P

JaagupAverin replied:
Yeah I agree this is too big of a hack around the actual issue so I think I'll remove this part for now and maybe open an issue for it instead.

JaagupAverin replied:
With the other adjustments my python project is 99.8% correctly highlighted though so its a good experience compared to a month ago already :)

JaagupAverin replied:
Also regarding semantic tokens I believe basedpyright already has support for this https://docs.basedpyright.com/latest/benefits-over-pyright/pylance-features/.
This would make it [another big feature](https://github.com/zed-industries/zed/issues/7296#issuecomment-2495505867) in support of basedpyright at this stage.

JaagupAverin committed:
python: Highlight decorators with arguments.

JaagupAverin committed:
python: Reorder highlighting priority.

Generic highlights should be higher up in the file in case they are conditionally overwritten by more specific rules below.

JaagupAverin committed:
python: Add attribute docstrings after type alias assignments.

These assignments are effectively a subset of expression_statements where `type X = Y` and are often documented using docstrings.

arv-anshul commented:
Hi @JaagupAverin, I have some issues around Syntax Highlighting too. Review and change it, if you find it right.

1. Mark these `not`, `in`, `is` operator as `operator.special`.

	<details>
	<summary>:camera: Screenshot</summary>
	
	![image](https://github.com/user-attachments/assets/1be854f6-ea96-4f8a-8a92-d57142aead48)
	
	</details>

2. Identify `# TODO:`, `# FIXME:` as special comment in comments (`comments.special`).
3. _(I don't know this)_; Can we identify function parameter as `function.parameter`.

**Thank You!**

JaagupAverin commented:
> Hi @JaagupAverin, I have some issues around Syntax Highlighting too. Review and change it, if you find it right.
> 
> 1. Mark these `not`, `in`, `is` operator as `operator.special`.
> 
> 	<details>
> 	<summary>:camera: Screenshot</summary>
> 	
> 	![image](https://github.com/user-attachments/assets/1be854f6-ea96-4f8a-8a92-d57142aead48)
> 	
> 	</details>
> 
> 2. Identify `# TODO:`, `# FIXME:` as special comment in comments (`comments.special`).
> 3. _(I don't know this)_; Can we identify function parameter as `function.parameter`.
> 
> **Thank You!**
> 

1. Personally I'm for extra configurability but I'm not sure if Zed team might be worried about polluting the grammar with very specific non standard rules? @osiewicz ? I think the current [list of captures](https://zed.dev/docs/extensions/languages#syntax-highlighting) is already out of date and I've thought if and where I should document the all the new captures for python? 
2. I really like this personally, would add it too.
3. This is doable in the function signature but not in the body without LSP highlighting support so I'm not sure if it would be very helpful as a half baked solution yet.

uncenter commented:
> I really like this personally, would add it too. Again I'd have to defer to Zed team if there is planned a more standard/documented way of approaching TODOs in code?

The way this is done elsewhere (nvim-treesitter) is by adding the following to `injections.scm`:
```scm
((comment) @injection.content
  (#set! injection.language "comment"))
```
And then https://github.com/stsewd/tree-sitter-comment (which syntax highlights those kind of TODO comments) is added under the "comment" language name, so injected for comments as such.

This currently isn't standard practice in Zed though. See https://github.com/zed-industries/zed/pull/9082#issuecomment-1997957426 and #11895.

uncenter commented:
> 1. Mark these `not`, `in`, `is` operator as `operator.special`.

I believe `keyword.operator` is what is normally used for these?

> 3. Can we identify function parameter as `function.parameter`.

I believe `variable.parameter` is what is standard for that.

JaagupAverin committed:
python: Add keyword.operator highlight.

arv-anshul commented:
Ok, Thanks for responding and correcting my comment @JaagupAverin and @uncenter.

Above @JaagupAverin mentioned about [semantic tokens highlighting through LSP](https://github.com/zed-industries/zed/pull/21389#discussion_r1865383237), specifically `basedpyright` (which has support for now). Is this feature is available now to use?

JaagupAverin commented:
I added one more commit for classifying keyword operators as `keyword.operator`.    
For #TODOs, I suppose #11895 would be a much better solution than a temporary hack.   
Also I had a look at the nvim implementation for implementing `@variable.parameter` and to my mind that is quite a [hardcore solution](https://github.com/nvim-treesitter/nvim-treesitter/blob/0c26a5c4bc407fdbca545ab25cf6084fd069444b/queries/python/highlights.scm#L114C1-L167C39) for something of minimal value (the parameters in declaration and body will be highlighted differently). Nvim has a few hacks that won't work correctly in all cases it seems.

uncenter commented:
> Nvim has a few hacks that won't work correctly in all cases it seems.

Can you elaborate? That list of queries for variable parameters seems fine to me fwiw.

JaagupAverin commented:
There's one where any 2nd argument to `isinstance` is always treated as type. Its somewhat giving special treatment to a function and might cause an identifier to switch colors back and forth (if it not properly identified as a `type` overall). There's a few other cases where best attempts are made to highlight types using their relations to keywords like `TypeAlias`, but the identifier will again just pop up with different colours throughout the file which imo is more confusing.
```
((call
  function: (identifier) @_isinstance
  arguments: (argument_list
    (_)
    (identifier) @type))
  (#eq? @_isinstance "isinstance"))
```
There's another that tries to fix [what I believe is a bug in the python tree-sitter](https://github.com/tree-sitter/tree-sitter-python/issues/291). This workaround won't always work since it could have several kinds of parents, not just `type`.
```
(type
  (subscript
    (identifier) @type)) ; type subscript: Tuple[int]
```
btw the only reason I picked these out was because I have been trying to solve the same issues here, but came to the conclusion that they're not worth the special handling. I thought there was another hack regarding "dotted" decorators, that I thought I had tried and had had issues with, but after trying it again it does seem to work :)

JaagupAverin committed:
python: Highlight decorators with "nesting".

osiewicz commented on crates/languages/src/python/highlights.scm:
```diff
@@ -47,16 +57,19 @@
    @function.builtin
    "^(abs|all|any|ascii|bin|bool|breakpoint|bytearray|bytes|callable|chr|classmethod|compile|complex|delattr|dict|dir|divmod|enumerate|eval|exec|filter|float|format|frozenset|getattr|globals|hasattr|hash|help|hex|id|input|int|isinstance|issubclass|iter|len|list|locals|map|max|memoryview|min|next|object|oct|open|ord|pow|print|property|range|repr|reversed|round|set|setattr|slice|sorted|staticmethod ...
 
+((identifier) @type.builtin
+    (#match? @type.builtin "^(int|float|complex|bool|list|tuple|range|str|bytes|bytearray|memoryview|set|frozenset|dict)$"))
+
```
```suggestion
((identifier) @type.builtin
    (#any-of? @type.builtin "int" "float" "complex" "bool" "list" "tuple" "range" "str" "bytes" "bytearray" "memoryview" "set" "frozenset" "dict"))
    
```

JaagupAverin committed:
Update crates/languages/src/python/highlights.scm

Co-authored-by: Piotr Osiewicz <24362066+osiewicz@users.noreply.github.com>

osiewicz commented:
Thanks!

osiewicz commented:
I've just noticed that there's a bunch of unresolved questions in this thread; I feel like these can be tackled in a separate PR.
